File: packages/sdk-js/src/framework-adapters.ts
import type { ResolutionRequest, ResolutionResponse } from "../../executor/src/index.js";
import type { AuthResolver, Caller, ExecutionEvent, RunLocalSkillResult } from "../../runner-local/src/index.js";

export interface FrameworkBridgeRunOptions {
  readonly skillPath: string;
  readonly inputs?: Readonly<Record<string, unknown>>;
  readonly answersPath?: string;
  readonly receiptDir?: string;
  readonly runxHome?: string;
  readonly parentReceipt?: string;
  readonly contextFrom?: readonly string[];
  readonly caller?: Caller;
  readonly authResolver?: AuthResolver;
  readonly allowedSourceTypes?: readonly string[];
  readonly resumeFromRunId?: string;
}

export type FrameworkSkillExecutor = (
  options: FrameworkBridgeRunOptions & { readonly caller: Caller },
) => Promise<RunLocalSkillResult>;

export interface FrameworkBoundaryContext {
  readonly request: ResolutionRequest;
  readonly events: readonly ExecutionEvent[];
}

export type FrameworkBoundaryReply =
  | ResolutionResponse
  | {
      readonly actor?: "agent" | "human";
      readonly payload: unknown;
    }
  | boolean
  | string
  | number
  | Readonly<Record<string, unknown>>
  | undefined;

export type FrameworkBoundaryResolver = (
  context: FrameworkBoundaryContext,
) => Promise<FrameworkBoundaryReply> | FrameworkBoundaryReply;

export interface FrameworkBridgeOptions {
  readonly execute: FrameworkSkillExecutor;
}

export interface FrameworkPausedResult {
  readonly status: "paused";
  readonly skillName: string;
  readonly runId: string;
  readonly requests: readonly ResolutionRequest[];
  readonly stepIds?: readonly string[];
  readonly stepLabels?: readonly string[];
  readonly events: readonly ExecutionEvent[];
}

export interface FrameworkCompletedResult {
  readonly status: "completed";
  readonly skillName: string;
  readonly receiptId: string;
  readonly output: string;
  readonly events: readonly ExecutionEvent[];
}

export interface FrameworkFailedResult {
  readonly status: "failed";
  readonly skillName: string;
  readonly receiptId?: string;
  readonly error: string;
  readonly events: readonly ExecutionEvent[];
}

export interface FrameworkDeniedResult {
  readonly status: "denied";
  readonly skillName: string;
  readonly reasons: readonly string[];
  readonly receiptId?: string;
  readonly events: readonly ExecutionEvent[];
}

export type FrameworkRunResult =
  | FrameworkPausedResult
  | FrameworkCompletedResult
  | FrameworkFailedResult
  | FrameworkDeniedResult;

export interface FrameworkBridge {
  readonly run: (
    options: FrameworkBridgeRunOptions & {
      readonly resolver?: FrameworkBoundaryResolver;
    },
  ) => Promise<FrameworkRunResult>;
  readonly resume: (
    runId: string,
    options: Omit<FrameworkBridgeRunOptions, "resumeFromRunId"> & {
      readonly resolver?: FrameworkBoundaryResolver;
    },
  ) => Promise<FrameworkRunResult>;
}

export interface OpenAIAdapterResponse {
  readonly role: "tool";
  readonly content: readonly [{ readonly type: "text"; readonly text: string }];
  readonly structuredContent: {
    readonly runx: FrameworkRunResult;
  };
}

export interface AnthropicAdapterResponse {
  readonly content: readonly [{ readonly type: "text"; readonly text: string }];
  readonly metadata: {
    readonly runx: FrameworkRunResult;
  };
}

export interface VercelAiAdapterResponse {
  readonly messages: readonly [{ readonly role: "assistant"; readonly content: string }];
  readonly data: {
    readonly runx: FrameworkRunResult;
  };
}

export interface LangChainAdapterResponse {
  readonly content: string;
  readonly additional_kwargs: {
    readonly runx: FrameworkRunResult;
  };
}

export interface CrewAiAdapterResponse {
  readonly raw: string;
  readonly json_dict: {
    readonly runx: FrameworkRunResult;
  };
}

export interface ProviderFrameworkAdapter<TResponse> {
  readonly run: (
    options: FrameworkBridgeRunOptions & {
      readonly resolver?: FrameworkBoundaryResolver;
    },
  ) => Promise<TResponse>;
  readonly resume: (
    runId: string,
    options: Omit<FrameworkBridgeRunOptions, "resumeFromRunId"> & {
      readonly resolver?: FrameworkBoundaryResolver;
    },
  ) => Promise<TResponse>;
}

export function createFrameworkBridge(options: FrameworkBridgeOptions): FrameworkBridge {
  const bridge: FrameworkBridge = {
    run: async (runOptions) => {
      const events: ExecutionEvent[] = [];
      const caller: Caller = {
        report: async (event) => {
          events.push(event);
          await runOptions.caller?.report(event);
        },
        resolve: async (request) => {
          const resolved = normalizeFrameworkReply(await runOptions.resolver?.({ request, events }), request);
          return resolved ?? await runOptions.caller?.resolve(request);
        },
      };

      const result = await options.execute({
        ...runOptions,
        caller,
      });
      return normalizeRunResult(result, events);
    },
    resume: async (runId, runOptions) => {
      return await bridge.run({
        ...runOptions,
        resumeFromRunId: runId,
      });
    },
  };
  return bridge;
}

export function createOpenAiAdapter(bridge: FrameworkBridge): ProviderFrameworkAdapter<OpenAIAdapterResponse> {
  return {
    run: async (options) => toOpenAiResponse(await bridge.run(options)),
    resume: async (runId, options) => toOpenAiResponse(await bridge.resume(runId, options)),
  };
}

export function createAnthropicAdapter(bridge: FrameworkBridge): ProviderFrameworkAdapter<AnthropicAdapterResponse> {
  return {
    run: async (options) => toAnthropicResponse(await bridge.run(options)),
    resume: async (runId, options) => toAnthropicResponse(await bridge.resume(runId, options)),
  };
}

export function createVercelAiAdapter(bridge: FrameworkBridge): ProviderFrameworkAdapter<VercelAiAdapterResponse> {
  return {
    run: async (options) => toVercelAiResponse(await bridge.run(options)),
    resume: async (runId, options) => toVercelAiResponse(await bridge.resume(runId, options)),
  };
}

export function createLangChainAdapter(bridge: FrameworkBridge): ProviderFrameworkAdapter<LangChainAdapterResponse> {
  return {
    run: async (options) => toLangChainResponse(await bridge.run(options)),
    resume: async (runId, options) => toLangChainResponse(await bridge.resume(runId, options)),
  };
}

export function createCrewAiAdapter(bridge: FrameworkBridge): ProviderFrameworkAdapter<CrewAiAdapterResponse> {
  return {
    run: async (options) => toCrewAiResponse(await bridge.run(options)),
    resume: async (runId, options) => toCrewAiResponse(await bridge.resume(runId, options)),
  };
}

function normalizeFrameworkReply(
  reply: FrameworkBoundaryReply,
  request: ResolutionRequest,
): ResolutionResponse | undefined {
  if (reply === undefined) {
    return undefined;
  }
  if (isResolutionResponse(reply)) {
    return reply;
  }
  if (typeof reply === "object" && reply !== null && "payload" in reply) {
    const candidate = reply as { readonly actor?: "agent" | "human"; readonly payload: unknown };
    return {
      actor: candidate.actor ?? defaultActorForRequest(request),
      payload: candidate.payload,
    };
  }
  if (typeof reply === "boolean" && request.kind === "approval") {
    return { actor: "human", payload: reply };
  }
  return {
    actor: defaultActorForRequest(request),
    payload: reply,
  };
}

function defaultActorForRequest(request: ResolutionRequest): "agent" | "human" {
  return request.kind === "cognitive_work" ? "agent" : "human";
}

function isResolutionResponse(value: unknown): value is ResolutionResponse {
  return typeof value === "object"
    && value !== null
    && "actor" in value
    && "payload" in value
    && (((value as { readonly actor?: unknown }).actor === "agent") || ((value as { readonly actor?: unknown }).actor === "human"));
}

function normalizeRunResult(result: RunLocalSkillResult, events: readonly ExecutionEvent[]): FrameworkRunResult {
  if (result.status === "needs_resolution") {
    return {
      status: "paused",
      skillName: result.skill.name,
      runId: result.runId,
      requests: result.requests,
      stepIds: result.stepIds,
      stepLabels: result.stepLabels,
      events,
    };
  }
  if (result.status === "policy_denied") {
    return {
      status: "denied",
      skillName: result.skill.name,
      reasons: result.reasons,
      receiptId: result.receipt?.id,
      events,
    };
  }
  if (result.status === "success") {
    return {
      status: "completed",
      skillName: result.skill.name,
      receiptId: result.receipt.id,
      output: result.execution.stdout,
      events,
    };
  }
  return {
    status: "failed",
    skillName: result.skill.name,
    receiptId: result.receipt.id,
    error: result.execution.errorMessage ?? (result.execution.stderr || result.execution.stdout),
    events,
  };
}

function toOpenAiResponse(result: FrameworkRunResult): OpenAIAdapterResponse {
  return {
    role: "tool",
    content: [{ type: "text", text: summarizeFrameworkResult(result) }],
    structuredContent: { runx: result },
  };
}

function toAnthropicResponse(result: FrameworkRunResult): AnthropicAdapterResponse {
  return {
    content: [{ type: "text", text: summarizeFrameworkResult(result) }],
    metadata: { runx: result },
  };
}

function toVercelAiResponse(result: FrameworkRunResult): VercelAiAdapterResponse {
  return {
    messages: [{ role: "assistant", content: summarizeFrameworkResult(result) }],
    data: { runx: result },
  };
}

function toLangChainResponse(result: FrameworkRunResult): LangChainAdapterResponse {
  return {
    content: summarizeFrameworkResult(result),
    additional_kwargs: { runx: result },
  };
}

function toCrewAiResponse(result: FrameworkRunResult): CrewAiAdapterResponse {
  return {
    raw: summarizeFrameworkResult(result),
    json_dict: { runx: result },
  };
}

function summarizeFrameworkResult(result: FrameworkRunResult): string {
  switch (result.status) {
    case "completed":
      return `${result.skillName} completed. Inspect receipt ${result.receiptId}.`;
    case "paused":
      return `${result.skillName} paused at ${result.runId}. Resume after resolving ${result.requests.length} request(s).`;
    case "denied":
      return `${result.skillName} was denied by policy.`;
    case "failed":
      return `${result.skillName} failed. Inspect receipt ${result.receiptId ?? "n/a"}.`;
  }
}

File: packages/sdk-python/runx/framework_adapters.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Callable, Mapping, Sequence

from . import RunxClient


@dataclass(frozen=True)
class FrameworkBoundaryContext:
    request: Mapping[str, Any]
    events: tuple[Mapping[str, Any], ...] = ()


FrameworkBoundaryResolver = Callable[[FrameworkBoundaryContext], Any | None]


@dataclass(frozen=True)
class FrameworkPausedResult:
    status: str
    skill_name: str
    run_id: str
    requests: tuple[Mapping[str, Any], ...]
    step_ids: tuple[str, ...] = ()
    step_labels: tuple[str, ...] = ()
    events: tuple[Mapping[str, Any], ...] = ()


@dataclass(frozen=True)
class FrameworkCompletedResult:
    status: str
    skill_name: str
    receipt_id: str
    output: str
    events: tuple[Mapping[str, Any], ...] = ()


@dataclass(frozen=True)
class FrameworkFailedResult:
    status: str
    skill_name: str
    error: str
    receipt_id: str | None = None
    events: tuple[Mapping[str, Any], ...] = ()


@dataclass(frozen=True)
class FrameworkDeniedResult:
    status: str
    skill_name: str
    reasons: tuple[str, ...]
    receipt_id: str | None = None
    events: tuple[Mapping[str, Any], ...] = ()


FrameworkRunResult = (
    FrameworkPausedResult
    | FrameworkCompletedResult
    | FrameworkFailedResult
    | FrameworkDeniedResult
)


class FrameworkBridge:
    def __init__(self, client: RunxClient) -> None:
        self.client = client

    def run(
        self,
        skill_path: str,
        inputs: Mapping[str, Any] | None = None,
        resolver: FrameworkBoundaryResolver | None = None,
    ) -> FrameworkRunResult:
        initial = self.client.run_skill(skill_path, inputs=inputs, non_interactive=True)
        return self._drive(initial, resolver=resolver)

    def resume(
        self,
        run_id: str,
        resolver: FrameworkBoundaryResolver | None = None,
    ) -> FrameworkRunResult:
        initial = self.client.resume_run(run_id)
        return self._drive(initial, resolver=resolver)

    def _drive(
        self,
        payload: Mapping[str, Any],
        resolver: FrameworkBoundaryResolver | None,
    ) -> FrameworkRunResult:
        current = dict(payload)
        while True:
            result = normalize_framework_result(current)
            if not isinstance(result, FrameworkPausedResult):
                return result
            if resolver is None:
                return result

            answers: dict[str, Any] = {}
            approvals: dict[str, bool] = {}
            for request in result.requests:
                reply = resolver(FrameworkBoundaryContext(request=request, events=result.events))
                normalized = _normalize_resolution_reply(request, reply)
                if normalized is None:
                    continue
                if request.get("kind") == "approval":
                    gate = request.get("gate") or {}
                    gate_id = str(gate.get("id") or "")
                    approvals[gate_id] = bool(normalized["payload"])
                    continue
                request_id = str(request.get("id") or "")
                answers[request_id] = normalized["payload"]

            if not answers and not approvals:
                return result

            current = self.client.resume_run(result.run_id, answers=answers, approvals=approvals)


class ProviderFrameworkAdapter:
    def __init__(self, bridge: FrameworkBridge, formatter: Callable[[FrameworkRunResult], Mapping[str, Any]]) -> None:
        self.bridge = bridge
        self.formatter = formatter

    def run(
        self,
        skill_path: str,
        inputs: Mapping[str, Any] | None = None,
        resolver: FrameworkBoundaryResolver | None = None,
    ) -> Mapping[str, Any]:
        return self.formatter(self.bridge.run(skill_path, inputs=inputs, resolver=resolver))

    def resume(
        self,
        run_id: str,
        resolver: FrameworkBoundaryResolver | None = None,
    ) -> Mapping[str, Any]:
        return self.formatter(self.bridge.resume(run_id, resolver=resolver))


def create_framework_bridge(client: RunxClient) -> FrameworkBridge:
    return FrameworkBridge(client)


def create_openai_adapter(bridge: FrameworkBridge) -> ProviderFrameworkAdapter:
    return ProviderFrameworkAdapter(bridge, _to_openai_response)


def create_anthropic_adapter(bridge: FrameworkBridge) -> ProviderFrameworkAdapter:
    return ProviderFrameworkAdapter(bridge, _to_anthropic_response)


def create_vercel_ai_adapter(bridge: FrameworkBridge) -> ProviderFrameworkAdapter:
    return ProviderFrameworkAdapter(bridge, _to_vercel_response)


def create_langchain_adapter(bridge: FrameworkBridge) -> ProviderFrameworkAdapter:
    return ProviderFrameworkAdapter(bridge, _to_langchain_response)


def create_crewai_adapter(bridge: FrameworkBridge) -> ProviderFrameworkAdapter:
    return ProviderFrameworkAdapter(bridge, _to_crewai_response)


def normalize_framework_result(payload: Mapping[str, Any]) -> FrameworkRunResult:
    status = str(payload.get("status") or "")
    skill = payload.get("skill")
    skill_name = str(skill.get("name")) if isinstance(skill, Mapping) else str(skill or "")
    if status == "needs_resolution":
        return FrameworkPausedResult(
            status="paused",
            skill_name=skill_name,
            run_id=str(payload.get("run_id") or ""),
            requests=tuple(payload.get("requests") or ()),
            step_ids=tuple(str(item) for item in payload.get("step_ids") or ()),
            step_labels=tuple(str(item) for item in payload.get("step_labels") or ()),
        )
    if status == "policy_denied":
        reasons = payload.get("reasons") or ()
        receipt = payload.get("receipt") or {}
        return FrameworkDeniedResult(
            status="denied",
            skill_name=skill_name,
            reasons=tuple(str(item) for item in reasons),
            receipt_id=_nested_str(receipt, "id"),
        )
    if status == "success":
        execution = payload.get("execution") or {}
        receipt = payload.get("receipt") or {}
        return FrameworkCompletedResult(
            status="completed",
            skill_name=skill_name,
            receipt_id=str(receipt.get("id") or ""),
            output=str(execution.get("stdout") or ""),
        )
    execution = payload.get("execution") or {}
    receipt = payload.get("receipt") or {}
    error = str(execution.get("errorMessage") or execution.get("stderr") or execution.get("stdout") or "")
    return FrameworkFailedResult(
        status="failed",
        skill_name=skill_name,
        error=error,
        receipt_id=_nested_str(receipt, "id"),
    )


def _normalize_resolution_reply(
    request: Mapping[str, Any],
    reply: Any | None,
) -> Mapping[str, Any] | None:
    if reply is None:
        return None
    if isinstance(reply, Mapping) and "actor" in reply and "payload" in reply:
        return {
            "actor": str(reply.get("actor") or _default_actor_for_request(request)),
            "payload": reply.get("payload"),
        }
    if isinstance(reply, Mapping) and "payload" in reply:
        return {
            "actor": str(reply.get("actor") or _default_actor_for_request(request)),
            "payload": reply.get("payload"),
        }
    if isinstance(reply, bool) and request.get("kind") == "approval":
        return {"actor": "human", "payload": reply}
    return {
        "actor": _default_actor_for_request(request),
        "payload": reply,
    }


def _default_actor_for_request(request: Mapping[str, Any]) -> str:
    return "agent" if request.get("kind") == "cognitive_work" else "human"


def _summary(result: FrameworkRunResult) -> str:
    if isinstance(result, FrameworkCompletedResult):
        return f"{result.skill_name} completed. Inspect receipt {result.receipt_id}."
    if isinstance(result, FrameworkPausedResult):
        return f"{result.skill_name} paused at {result.run_id}. Resume after resolving {len(result.requests)} request(s)."
    if isinstance(result, FrameworkDeniedResult):
        return f"{result.skill_name} was denied by policy."
    return f"{result.skill_name} failed. Inspect receipt {result.receipt_id or 'n/a'}."


def _to_openai_response(result: FrameworkRunResult) -> Mapping[str, Any]:
    return {
        "role": "tool",
        "content": [{"type": "text", "text": _summary(result)}],
        "structuredContent": {"runx": _result_to_dict(result)},
    }


def _to_anthropic_response(result: FrameworkRunResult) -> Mapping[str, Any]:
    return {
        "content": [{"type": "text", "text": _summary(result)}],
        "metadata": {"runx": _result_to_dict(result)},
    }


def _to_vercel_response(result: FrameworkRunResult) -> Mapping[str, Any]:
    return {
        "messages": [{"role": "assistant", "content": _summary(result)}],
        "data": {"runx": _result_to_dict(result)},
    }


def _to_langchain_response(result: FrameworkRunResult) -> Mapping[str, Any]:
    return {
        "content": _summary(result),
        "additional_kwargs": {"runx": _result_to_dict(result)},
    }


def _to_crewai_response(result: FrameworkRunResult) -> Mapping[str, Any]:
    return {
        "raw": _summary(result),
        "json_dict": {"runx": _result_to_dict(result)},
    }


def _result_to_dict(result: FrameworkRunResult) -> Mapping[str, Any]:
    if isinstance(result, FrameworkPausedResult):
        return {
            "status": result.status,
            "skill_name": result.skill_name,
            "run_id": result.run_id,
            "requests": list(result.requests),
            "step_ids": list(result.step_ids),
            "step_labels": list(result.step_labels),
            "events": list(result.events),
        }
    if isinstance(result, FrameworkCompletedResult):
        return {
            "status": result.status,
            "skill_name": result.skill_name,
            "receipt_id": result.receipt_id,
            "output": result.output,
            "events": list(result.events),
        }
    if isinstance(result, FrameworkDeniedResult):
        return {
            "status": result.status,
            "skill_name": result.skill_name,
            "reasons": list(result.reasons),
            "receipt_id": result.receipt_id,
            "events": list(result.events),
        }
    return {
        "status": result.status,
        "skill_name": result.skill_name,
        "error": result.error,
        "receipt_id": result.receipt_id,
        "events": list(result.events),
    }


def _nested_str(record: Mapping[str, Any], key: str) -> str | None:
    value = record.get(key)
    return None if value is None else str(value)
